src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisState.PerAnalyzerState.cs

```diff
@@ -157,12 +157,17 @@ public bool HasPendingSymbolAnalysis(ISymbol symbol)
             {
                 if (pendingEntities.TryGetValue(analysisEntity, out var state))
                 {
+                    if (state != null && state.StateKind == StateKind.InProcess)
```

Comment:
>if (state != null && state.StateKind == StateKind.InProcess) [](http://example.com/codeflow?start=20&length=60)

Are we always under lock here? Can other thread change state to ```InProcess``` before we complete operations below?

Reply:
Yes, we are always under lock here. Changing the state to `InProcess` can only happen under lock, so it cannot happen simultaneously here.